server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNode.java

```diff
@@ -190,10 +203,27 @@ public static DesiredNode readFrom(StreamInput in) throws IOException {
         }
         final var memory = ByteSizeValue.readFrom(in);
         final var storage = ByteSizeValue.readFrom(in);
-        final var version = Version.readVersion(in);
+        final String version;
+        if (in.getTransportVersion().onOrAfter(TransportVersions.DESIRED_NODE_VERSION_OPTIONAL_STRING)) {
+            version = in.readOptionalString();
+        } else {
+            version = Version.readVersion(in).toString();
+        }
         return new DesiredNode(settings, processors, processorsRange, memory, storage, version);
     }
 
+    private static final Pattern SEMANTIC_VERSION_PATTERN = Pattern.compile("^(\\d+\\.\\d+\\.\\d+)\\D?.*");
+
+    private static Version parseLegacyVersion(String version) {
+        if (version != null) {
+            var semanticVersionMatcher = SEMANTIC_VERSION_PATTERN.matcher(version);
```

Comment:
Why do we need this check?

Reply:
Since the API now accepts anything as `node_version`, there is change this is an empty string or anything really. An alternative is to have this check during parsing (line 108), so we keep rejecting invalid/non-semantic versions (you either skip it or provide a valid one). 
We don't really need it, we just need a Version if we are talking to older nodes (TransportVersion < DESIRED_NODE_VERSION_OPTIONAL_STRING), that's why the check is here.

Reply:
But let me know if you prefer to move it in the parser @fcofdez

Reply:
I think it's easier to understand if we keep the validation in the parser, but it's not a strong preference 👍